### ISSUE TITLE
256 template variant

### ISSUE THREAD

**Issue by JamyGolden** [mk12/base16-kitty#1]
Hey @mk12, I have a question which you may not know the answer to, but I think you might, but first a bit of context:

I'm an admin and maintainer of https://github.com/tinted-theming which you briefly looked into in a bat-cli issue several months ago.

I'm currently creating a https://github.com/tinted-theming terminal repo which consolidates all our terminal repos into a single repo to make it easier to maintain. Most terminal templates are pretty straight forward and don't change since it's just about setting the ANSI colors.

In creating this consolidated terminal repo (still private atm), I've removed the old "16 variant" template themes and renamed the `*-256*` variant themes to be the default, like you have done with your fork of base16-kitty, however I've gotten some push back on this, but there isn't any recollection as to why the "16 variant" templates exist in the first place and apparently there was a lot of discussion about this on Chris' old repos, but the issues can't be found on web.archive.org.

Since I'm the most active admin and maintainer at the moment, I don't feel comfortable keeping the "16 variant" templates without being able to explain why they are there. At a guess, it seems like people thought "ANSI has 16 main colors, base16 has 16 colors so let's map out all base16 to ANSI". Back then a styling pattern and [styling guidelines](https://github.com/tinted-theming/home/blob/main/styling.md#specific-colours-and-their-usages) didn't really exist so I can see people using any color they wanted for base00-base07 (generally a gradient now from ANSI0 to ANSI7) and base08, base0A-base0E (generally ANSI 1-6) as suggested in the https://github.com/tinted-theming/home/blob/main/styling.md#specific-colours-and-their-usages

On top of the above, I don't quite understand the need for shimming [tinted-shell](https://github.com/tinted-theming/tinted-shell) (base16-shell renamed since it supports multiple scheme systems now), aside from shimming colors outside ANSI 16 if the terminal doesn't support setting them directly in the theme.

There has just been so much talk about base16 16 variant and 256 variant terminal themes, but I can't see why we don't just duplicate the colors of the base16 ANSI bright colors with the standard color variants and I don't see how tinted-shell contributes anything meaningful for any terminal that can set ANSI 16-21 in the theme itself. I'd really appreciate if you can share your thoughts and any memories of this context if you have any :smile:

**Comment by mk12** [comment]
Hey @JamyGolden - sure, happy to help.

Indeed both systems (base16 and 4-bit ansi color) have a palette of 16 colors. The first 8 ansi colors are 0=black, 1=red, 2=green, 3=yellow, 4=blue, 5=purple, 6=cyan, and 7=white. Here things work pretty well: slots 1-6 match 6 of base16's 8 colors, leaving out orange and brown. We have to reinterpret 0 as background and 7 as foreground. When terminal software does this, it plays nicely with switching between light and dark modes. (If you want this property _without_ relying on the base16 interpretation of 0 and 7, you must leave bg/fg unset when you want the theme's bg/fg. When your CLI only needs 8 colors, it's best to use this approach for maximum compatibility.)

Then we get to the other 8. In the ANSI palette they are the "bright" colors. (Some terminals are configured to also use bold for bright colors.) Sticking with 0=bg and 7=fg, there's a good interpretation for 8 and 15. Since 0 on 0 = bg on bg is invisible, 8 = "bright 0" on 0 should be visible but faded — the shade of gray used for comments, base03. Similarly since 7 on 0 is the default, 15 = "bright 7" on 0 should be even higher contrast — that's base07.

The problem now is with the remaining 6 shades of grey. There are three options:

1. Put them in bright color slots arbitrarily.
    - **Pros**: base16-configured programs can use them.
    - **Cons**: programs that are trying to use "bright" colors will look wrong or even unreadable.
2. Don't use them at all. Just copy over the non-bright colors to the remaining 6 bright slots.
    - **Pros**: avoids the cons of (1).
    - **Cons**: programs configured with base16 themes are only able to use 10 colors from the palette.
3. Like (2), but put the remaining colors in slots 16-21.
    - **Pros**: avoids the cons of (1) and programs can still use the full palette. There is little issue in practice with other uses of 16-21 since programs nowadays would use 24-bit color instead.
    - **Cons**: not all terminals let you configure 16-255 from the config file.

In base16 theme repos, the regular files followed (1) and the "256" variants followed (2). You were then meant to achieve (3) by using base16-shell, which would set 16-21 using escape codes. This always felt like a terrible hack to me. It defeats the purpose of base16, because those escape codes only apply to your current shell session. Doing (3) properly by configuring 16-21 globally in your terminal means that when you change the theme, everything will update at once, including existing output in the scroll back, even if nested within multiple ssh/tmux sessions.

My recommendation is to continue offering both variants for terminal programs (e.g. vim), but only offer one or the other for terminals. Maybe rename them so that (3) is the normal one and (1) has a "-16" suffix. For example:

- macOS Terminal.app: `tinted-16.terminal`
- kitty: `tinted.conf`
- vim: `tinted.vim` and `tinted-16.vim` — use the one that matches your terminal

Then you could tell people to beware the "-16" versions will sometimes result in incorrect/undreadable colors if other programs think they are the bright colors.

**Comment by JamyGolden** [comment]
Thanks for the detailed response, I really appreciate it! 

I think a big part I was missing was using the `-16` terminal theme variants with `-16` cli template themes, for example https://github.com/tinted-theming/tinted-fzf/blob/main/ansi/ansi.sh looks broken with `-16` terminal themes since it uses bright ANSI colors and creating a `fzf/ansi-16.sh` would sort that out. iirc hat fzf ANSI theme is the only theme we have that actually uses ANSI colors, the rest of our template themes "hardcode" the hex/rgb/etc values (most of the template themes are for gui applications though).

I need to go through the tinted-vim theme again, but we use a builder and "hardcode" the values in there too, so someone isn't required to use a base16 terminal theme to use the tinted-vim themes. Do you use ANSI passthrough themes for all your cli tools, including vim?

**Comment by mk12** [comment]
You're right that if you configure terminal programs to use 24-bit color and hardcode the RRGGBB then none of this matters. I forgot about that. I like having a level of indirection through the terminal's palette because then existing output updates when you change themes and ssh/tmux just works. I only ever used base16 in the terminal, not in GUIs, so I might care about this more than other people. Last time I looked, the vim plugin only hardcodes the colors when `termguicolors` is set, and otherwise does the 16/256 thing. I do `set notermguicolors` in my vimrc on purpose.

Here's another idea then, given a theme "foo":

- Terminals:
    - Offer `foo` which does (2), i.e. 8=base03, 9...14=1...6, 15=base07.
    - If the terminal can configure the 256 palette, additionally offer `foo-256` which does (3).
- Terminal programs:
    - Offer `256` which does (3).
    - If the program can emit 24-bit color scapes, offer `foo` which hardcodes colors.

For example:

- macOS Terminal.app: `foo`, `bar`, `baz`
- kitty: `foo`, `foo-256`, `bar`, `bar-256`, `baz`, `baz-256`
- vim: `256`, `foo`, `bar`, `baz`
- program-that-cant-do-24-bit-color: `256`

The docs would then be simple for most people. If you want "foo", just use "foo" everywhere. For programs that can't be configured with 24-bit color (note fzf is not an example, it supports #rrggbb), there is no need to apply a theme. They will just use the terminal's ansi colors.

For users that switch themes often and care about the existing output/ssh/tmux stuff, they can (1) choose a terminal that offers "-256" themes, (2) pick one of those themes, and (3) use the "256" theme in all their terminal programs.

**Comment by JamyGolden** [comment]
Do you see a reason not to offer (3) (your suggested `foo-256`) as the default so named `foo`, provided the terminal supports setting extended ANSI colors and then (2) named `foo-16` if the terminals don't support it (seems like Kermit and Rio are the 2 terminals we support that don't support setting extended colors).

**Comment by mk12** [comment]
Actually yes that’s simpler. Just have one file for terminal themes, which is (3) if possible else (2). Then terminal users who want the palette indirection can use “256” instead of a specific named theme for all their terminal programs. You can warn that ”256” won’t work if using Kermit or Rio (16-21 will probably be xterm default shades of blue). I thought macOS Terminal.app also didn’t support configuring colors 16-255 but I could be wrong.

**Comment by JamyGolden** [comment]
Yeah you're right about MacOS terminal not supporting 256.

Could you have a look at https://github.com/tinted-theming/tinted-terminal? I've put a summary of what we've spoken about under the "Background" subheading. Do you think I've left anything out or any other general suggestions? I've sent you an invite to a general purpose tinted-theming team which should give you read access to the above private repo.

I've added a thanks to you at the bottom of the readme. I can edit or remove if you'd prefer it not be there (I can rebase it out).

**Comment by mk12** [comment]
Yes, that looks fine to me. This is the first I've seen of "base 24" though. It sounds like this is to theme the bright colors as well? I have a few thoughts on that:

- Why not make base24 the only option? Is there any situation where someone would _not_ want to be able to distinguish the bright colors?
- It would just be base22 if you make ansi 8 = bright black = base03 and ansi 15 = bright white = base07 as discussed earlier. However, if you can make all the themes have distinct bright colors, maybe there's no need to worry about treating ansi 8 and 15 specially, and you could actually do base24.
- I think https://github.com/tinted-theming/base24/blob/main/styling.md is incorrect: it maps ansi 0 to base01 and ansi 8 to base02, whereas I think they should be base00 (bg) and base03 (comment) respectively.

**Comment by JamyGolden** [comment]
Yeah I agree Base22 makes more sense, the extra 2 colors for darker and darkest is superfluous imo, but it's got the `24` in it's name though and `base10` and `base11` are the "dark" colors so it would require an index shift for the bright colors if we used a "Base22"

- Base24 would be a better option, over Base16, for the theme user, however Base16 means something to lot more to people than Base24, so reluctance to adopt something new (whether that be theme users or template maintainers). 

- If Base16 template maintainers used our [schemes](https://github.com/tinted-theming/schemes) repo, they could still switch to a base24 builder and build base16 themes if that's what they wanted, but communicating this, and any change, is difficult since the only way would be to go to the different base16 template repos and information in an issue.

- Those ANSI colors mapped 0 to base01 and 8 to base02 come from base24 originally but we've put this onto the base16 styling.md doc too. It should first be changed there and then I'll bring the changes into Base24: https://github.com/tinted-theming/home/pull/123 contains a PR

At least these are the reasons we haven't done it yet, not pushing back against the idea in general; Base24 was created as a superset of Base16 initially to be a replacement. The Base24 repos were transferred to tinted-theming about 2 years ago and we didn't want to make a big change by replacing Base16 but may we can start looking at that and maybe we create Base22 with a name not tied to the number of colors haha.

Edit: There's been talk of creating an ANSI16 scheme specification. I think this is probably the way to go. Template builders could generate some missing grays/gradients for UI templates if people wanted to use that scheme system for a UI, with these "grays" being generated (based on Black, Bright Black, White and Bright White) it would put this on par with Base22 for UI templates conceptually. But something like `Base24` or `BaseX` would be easier for people switch to since it wound't require them to change their templates when building with the new scheme-system.

**Comment by JamyGolden** [comment]
Outside of what we've been talking about: Would you be interested in being the maintainer of tinted-terminal's kitty and/or being more deeply involved with the admins and the direction of tinted-theming? You have a lot of base16 institutional knowledge as well as terminal and theming passion so I thought I'd ask at least :sweat_smile:
